File: internal/model/token.go
package model

import (
	"time"

	"github.com/usesnipet/snipet/pkg/jsonx"
)

type TokenType string

const (
	TokenTypeRefresh         TokenType = "refresh"
	TokenTypeActivateAccount TokenType = "activate_account"
	TokenTypeResetPassword   TokenType = "reset_password"
)

type Token struct {
	ID string `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`

	Type      TokenType     `gorm:"type:varchar(255);not null;index" json:"type"`
	Hash      string        `gorm:"type:varchar(255);not null;uniqueIndex" json:"-"`
	UserID    string        `gorm:"type:uuid;not null;index" json:"user_id"`
	ExpiresAt time.Time     `gorm:"type:timestamp;not null" json:"expires_at"`
	RevokedAt *time.Time    `gorm:"type:timestamp" json:"revoked_at"`
	Metadata  jsonx.JSONMap `gorm:"type:jsonb;not null" json:"metadata"`

	CreatedAt time.Time `gorm:"type:timestamp;not null;default:now()" json:"created_at"`

	User User `gorm:"foreignKey:UserID;references:ID;constraint:OnDelete:CASCADE" json:"-"`
}

File: web/src/models/token.ts
import { z } from "zod";

import { userSchema, type User } from "@/models/user";

export const tokenTypeSchema = z.enum(["refresh", "activate_account", "reset_password"]);
export type TokenType = z.infer<typeof tokenTypeSchema>;

export interface Token {
  id: string;
  type: TokenType;
  user_id: string;
  expires_at: Date;
  revoked_at: Date | null;
  metadata: Record<string, unknown>;
  created_at: Date;
  user?: User | null;
}

export const tokenSchema: z.ZodType<Token> = z.lazy(() =>
  z
    .object({
      id: z.uuid(),
      type: tokenTypeSchema,
      user_id: z.uuid(),
      expires_at: z.coerce.date(),
      revoked_at: z.coerce.date().nullable(),
      metadata: z.record(z.string(), z.unknown()),
      created_at: z.coerce.date(),
      user: userSchema.nullable().optional(),
    })
    .strict(),
);
